src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs

```diff
@@ -12538,16 +12550,31 @@ internal bool TryReadPlpUnicodeChars(ref char[] buff, int offst, int len, TdsPar
                 return true;       // No data
             }
 
-            Debug.Assert(((ulong)stateObj._longlen != TdsEnums.SQL_PLP_NULL),
-                    "Out of sync plp read request");
+            Debug.Assert(((ulong)stateObj._longlen != TdsEnums.SQL_PLP_NULL),"Out of sync plp read request");
 
             Debug.Assert((buff == null && offst == 0) || (buff.Length >= offst + len), "Invalid length sent to ReadPlpUnicodeChars()!");
             charsLeft = len;
 
-            // If total length is known up front, allocate the whole buffer in one shot instead of realloc'ing and copying over each time
-            if (buff == null && stateObj._longlen != TdsEnums.SQL_PLP_UNKNOWNLEN)
+            // If total length is known up front, the length isn't specified as unknown 
+            // and the caller doesn't pass int.max/2 indicating that it doesn't know the length
+            // allocate the whole buffer in one shot instead of realloc'ing and copying over each time
+            if (buff == null && stateObj._longlen != TdsEnums.SQL_PLP_UNKNOWNLEN && len < (int.MaxValue >> 1))
             {
-                buff = new char[(int)Math.Min((int)stateObj._longlen, len)];
+                if (supportRentedBuff && len < 1073741824) // 1 Gib
+                {
+                    if (buff != null && rentedBuff)
+                    {
+                        buff.AsSpan(0, offst).Clear();
+                        ArrayPool<char>.Shared.Return(buff, clearArray: false);
+                    }
```

Comment:
`buff` is always `null` here
```suggestion
```

Reply:
It isn't. If the pre loop check finds that that length was passed in as unknown and the caller hasn't provided a buffer then it will hit this piece of code and then rent/allocate a correctly sized buffer once it has calculated the required size. 

This is now the desired path because otherwise the pre-loop rental/allocate will create a `char[int.MaxValue/2]` size buffer which is usually far far too large and not something we want to keep around in the array pool because it will contribute to memory pressure.

Reply:
Hm not sure if I am missing something but two lines above there is this check
`if (buff == null && stateObj._longlen != TdsEnums.SQL_PLP_UNKNOWNLEN && len < (int.MaxValue >> 1))`
which means that `buff` can't be `null` here or?
Or is the `buff == null` check wrong?

Reply:
Ah, my mistake, I thought this was about the loop version. It's the preloop version. Yes, looks like it no longer applies,